src/map/achievement.c

```diff
@@ -0,0 +1,1693 @@
+	}
+
+	return a;
+}
+
+/**
+ * Calculates the achievement's totals via reference.
+ * @param sd         pointer to map_session_data
+ * @param points     pointer to total points var
+ * @param total      pointer to total var
+ * @param completed  pointer to completed var
+ * @param rank       pointer to achievement rank var
+ */
+void achievement_calculate_totals(const struct map_session_data *sd, int *total_points, int *completed, int *rank, int *curr_rank_points)
+{
+	const struct achievement *a = NULL;
+	const struct achievement_data *ad = NULL;
+	int i = 0;
+
+	nullpo_retv(sd);
+	
+	for (i = 0; i < VECTOR_LENGTH(sd->achievement); i++) {
+		a = &VECTOR_INDEX(sd->achievement, i);
+
+		if ((ad = achievement->get(a->id)) == NULL)
+			continue;
+
+		if (a->completed_at != 0 && (total_points != NULL || completed != NULL)) {
+			if (total_points != NULL)
+				*total_points += ad->points;
+			if (completed != NULL)
+				(*completed)++;
+		}
+	}
+
+	if (total_points != NULL && *total_points > 0
+		&& (rank != NULL || curr_rank_points != NULL)) {
+		if (curr_rank_points != NULL )
+			*curr_rank_points = *total_points;
+		for (i = 0; i < MAX_ACHIEVEMENT_RANKS && *curr_rank_points >= VECTOR_INDEX(achievement->rank_exp, i) && i < VECTOR_LENGTH(achievement->rank_exp); i++) {
+			if (curr_rank_points != NULL)
+				*curr_rank_points -= VECTOR_INDEX(achievement->rank_exp, i);
+			if (rank != NULL)
+				(*rank)++;
+		}
+	} else {
+		if (curr_rank_points != NULL) *curr_rank_points = 0;
+		if (rank != NULL) *rank = 0;
+	}
+}
+
+/**
+ * Checks whether all objectives of the achievement are completed.
+ * @param sd    as the map_session_data pointer
+ * @param ad    as the achievement_data pointer
+ * @return true if complete, false if not.
+ */
+bool achievement_check_complete(struct map_session_data *sd, struct achievement_data *ad)
+{
+	struct achievement *ach = NULL;
+	int i;
+
+	nullpo_retr(false, sd);
+	nullpo_retr(false, ad);
+
+	if ((ach = achievement->find_or_create(sd, ad->id, false)) == NULL)
+		return false;
+
+	for (i = 0; i < VECTOR_LENGTH(ad->objective); i++)
+		if (ach->objective[i] < VECTOR_INDEX(ad->objective, i).goal)
+			return false;
+
+	return true;
+}
+
+/**
+ * Compares the progress of an objective against it's goal.
+ * Increments the progress of the objective by the specified amount, towards the goal.
+ * @param sd      as a pointer to map_session_data
+ * @param ad      as a pointer to the achievement_data
+ * @param obj_idx as the index of the objective.
+ * @param progress   as the progress of the objective to be added.
+ */
+void achievement_progress_add(struct map_session_data *sd, struct achievement_data *ad, const unsigned int obj_idx, const int progress)
+{
+	struct achievement *ach = NULL;
+
+	nullpo_retv(sd);
+	nullpo_retv(ad);
+
+	Assert_retv(progress > 0);
+	Assert_retv(obj_idx < VECTOR_LENGTH(ad->objective));
+
+	// find existing or create a new entry.
+	ach = achievement->find_or_create(sd, ad->id, true);
```

Comment:
But it is exactly line 72 that I'm speaking about. If the id is invalid, line 72 causes this function to return NULL